src/server/pkg/storage/fileset/merge.go

```diff
@@ -31,64 +31,80 @@ func newMergeReader(rs []*Reader) *MergeReader {
 	return &MergeReader{pq: newPriorityQueue(fileStreams)}
 }
 
-func (mr *MergeReader) iterate(f func(*FileMergeReader) error) error {
+func (mr *MergeReader) iterate(cb func(*FileMergeReader) error) error {
 	return mr.pq.iterate(func(ss []stream, _ ...string) error {
-		// Convert generic streams to file streams.
-		var fss []*fileStream
-		for _, s := range ss {
-			fss = append(fss, s.(*fileStream))
-		}
-		// Create file merge reader and execute callback.
-		var frs []*FileReader
-		for _, fs := range fss {
-			fr, err := fs.r.Next()
-			if err != nil {
-				return err
-			}
-			frs = append(frs, fr)
-		}
-		frs = applyFullDeletes(frs)
-		contentTags := computeContentTags(frs)
-		if len(contentTags) == 0 {
-			return nil
-		}
-		return f(newFileMergeReader(frs, sizeOfTags(contentTags)))
+		return mr.mergeFile(ss, cb)
 	})
 }
 
-func applyFullDeletes(frs []*FileReader) []*FileReader {
-	var result []*FileReader
-	// Filter out the file readers with lower priority than the highest priority file reader that contains a full file deletion.
-	for _, fr := range frs {
-		result = append(result, fr)
+func (mr *MergeReader) mergeFile(ss []stream, cb func(*FileMergeReader) error, deleter ...func(string, ...string)) (retErr error) {
+	// Convert generic streams to file streams.
+	var fss []*fileStream
+	for _, s := range ss {
+		fss = append(fss, s.(*fileStream))
+	}
+	// Progress each stream and collect each file reader.
+	var i int
+	var frs []*FileReader
+	for ; i < len(fss); i++ {
+		fr, err := fss[i].r.Next()
+		if err != nil {
+			return err
+		}
+		frs = append(frs, fr)
+		// Filter out the file readers with lower priority than the highest priority file reader that contains a full file deletion.
 		if isFullDelete(fr.Index()) {
-			return result
+			break
+		}
+	}
+	// Progress each lower priority stream past the deleted file(s).
+	for i += 1; i < len(fss); i++ {
+		fr, err := fss[i].r.Next()
+		if err != nil {
+			return err
+		}
+		// If the full delete is for a directory, progress all lower priority streams past the directory.
+		if strings.HasSuffix(fr.Index().Path, "/") {
+			next := strings.TrimRight(fr.Index().Path, "/") + "0"
+			if err := fss[i].r.iterate(func(_ *FileReader) error { return nil }, next); err != nil {
+				return err
+			}
 		}
 	}
-	return result
+	contentTags, deleteTags := computeTags(frs)
+	if len(contentTags) == 0 {
+		if len(deleter) > 0 && len(deleteTags) > 0 {
```

Comment:
Is deleter just a predicate? I think this would be more clear as a non variadic argument.   You could set it to always return true if it was nil.

Reply:
No, deleter is a callback that performs a deletion. It is optional functionality provided by `mergeFile` and `mergeTag`.

Reply:
Regardless of how you want to encode the parameter I still think it makes the core algorithm more clear to deal with the peculiarities of whatever parameter encoding once at the top, and turn those branches into calls to a no-op.